### ISSUE TITLE
Calling `Deno.mainModule` causes Deno to panic using standalone/compiled Windows binary built on Linux

### ISSUE THREAD

**Issue by arite** [denoland/deno#19905]
## Overview

Deno panics when calling `Deno.mainModule` on Windows using a binary created on Linux with `deno compile --target x86_64-pc-windows-msvc`.

This is because the [`ModuleSpecifier`](https://github.com/denoland/deno_core/blob/0.195.0/core/module_specifier.rs#L52) (a [`Url`](https://docs.rs/url/latest/url/struct.Url.html)), which in this case is a URL for a Unix file path (e.g. `file:///tmp/mainModuleExample.js`), cannot be converted to a Windows file path when [`to_file_path()`](https://docs.rs/url/latest/url/struct.Url.html#method.to_file_path) is called. This result is unwrapped in `op_main_module` [here](https://github.com/denoland/deno/blob/v1.35.2/runtime/ops/runtime.rs#L23) which causes a panic.

This issue was reproduced using **Deno v1.35.2**.

Related issue: #7315

## Steps to Reproduce

On Linux (or equivalent), given a file name `/tmp/mainModuleExample.js`:
```js
console.log("Deno.mainModule:", Deno.mainModule);
```
Compiling and then running produces:
```bash
$ deno compile --allow-read /tmp/mainModuleExample.js
Compile file:///tmp/mainModuleExample.js to mainModuleExample

$ ./mainModuleExample 
Deno.mainModule: file:///tmp/mainModuleExample.js
```
Changing the target to Windows using `--target x86_64-pc-windows-msvc`:
```bash
$ deno compile --target x86_64-pc-windows-msvc --allow-read /tmp/mainModuleExample.js
Compile file:///tmp/mainModuleExample.js to mainModuleExample.exe
Archive:  /tmp/.tmpISnQkU/deno.zip
  inflating: deno.exe         
```
Results in the following when running on Windows:
```
>mainModuleExample.exe

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: windows x86_64
Version: 1.35.2
Args: ["mainModuleExample.exe"]

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', runtime\ops\runtime.rs:23:45
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Re-running with `RUST_BACKTRACE=full`:
```
>mainModuleExample.exe

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: windows x86_64
Version: 1.35.2
Args: ["mainModuleExample.exe"]

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', runtime\ops\runtime.rs:23:45
stack backtrace:
   0:     0x7ff6c6c41630 - napi_release_threadsafe_function
   1:     0x7ff6c5d056db - napi_release_threadsafe_function
   2:     0x7ff6c6c1ec89 - napi_release_threadsafe_function
   3:     0x7ff6c6c4596b - napi_release_threadsafe_function
   4:     0x7ff6c6c4578f - napi_release_threadsafe_function
   5:     0x7ff6c6c45380 - napi_release_threadsafe_function
   6:     0x7ff6c5b2146b - napi_fatal_error
   7:     0x7ff6c6c461f9 - napi_release_threadsafe_function
   8:     0x7ff6c6c45f95 - napi_release_threadsafe_function
   9:     0x7ff6c6c45ee9 - napi_release_threadsafe_function
  10:     0x7ff6c6c45ed4 - napi_release_threadsafe_function
  11:     0x7ff6c7ead5d5 - CrashForExceptionInNonABICompliantCodeRange
  12:     0x7ff6c7eadaf3 - CrashForExceptionInNonABICompliantCodeRange
  13:     0x7ff6c6618151 - napi_release_threadsafe_function
  14:     0x7ff6c7bb985e - CrashForExceptionInNonABICompliantCodeRange
  15:     0x7ff6c7bb801b - CrashForExceptionInNonABICompliantCodeRange
  16:     0x7ff6c7bc51ca - CrashForExceptionInNonABICompliantCodeRange
  17:     0x7ff6c7d12d7e - CrashForExceptionInNonABICompliantCodeRange
  18:     0x7ff6c7bb801b - CrashForExceptionInNonABICompliantCodeRange
  19:     0x7ff6c7bf734d - CrashForExceptionInNonABICompliantCodeRange
  20:     0x7ff6c7bb605c - CrashForExceptionInNonABICompliantCodeRange
  21:     0x7ff6c7bb5c57 - CrashForExceptionInNonABICompliantCodeRange
  22:     0x7ff6c6fec1f8 - napi_release_threadsafe_function
  23:     0x7ff6c6fecd45 - napi_release_threadsafe_function
  24:     0x7ff6c6feceed - napi_release_threadsafe_function
  25:     0x7ff6c6ff4f35 - napi_release_threadsafe_function
  26:     0x7ff6c6ff4621 - napi_release_threadsafe_function
  27:     0x7ff6c6ff4064 - napi_release_threadsafe_function
  28:     0x7ff6c6fef294 - napi_release_threadsafe_function
  29:     0x7ff6c6ea2e21 - napi_release_threadsafe_function
  30:     0x7ff6c6e92de7 - napi_release_threadsafe_function
  31:     0x7ff6c5ec6575 - napi_release_threadsafe_function
  32:     0x7ff6c5b05a1a - napi_fatal_error
  33:     0x7ff6c5b01cf5 - napi_fatal_error
  34:     0x7ff6c5b001b7 - napi_fatal_error
  35:     0x7ff6c5a6e15e - napi_fatal_error
  36:     0x7ff6c58959b4 - napi_cancel_async_work
  37:     0x7ff6c56dd85f - <unknown>
  38:     0x7ff6c5c06585 - napi_release_threadsafe_function
  39:     0x7ff6c57c2ed2 - napi_cancel_async_work
  40:     0x7ff6c5b2346e - napi_fatal_error
  41:     0x7ff6c56b40be - <unknown>
  42:     0x7ff6c5c0669e - napi_release_threadsafe_function
  43:     0x7ff6c7e5faec - CrashForExceptionInNonABICompliantCodeRange
  44:     0x7ff9d8977614 - BaseThreadInitThunk
  45:     0x7ff9d9a026b1 - RtlUserThreadStart

```
## Candidate Solutions

The `ModuleSpecifier` is converted to a file path in order to perform read permission check on the path. In this case of a `deno compile`d program the ModuleSpecifier represents the file path used when `deno compile` was invoked. As this was done at compile-time the runtime check does not seem necessary as the file path may no longer exist/could be from a different system or not be a valid platform-specific file path.

Options:

1. Remove permissions check for `Deno.mainModule` as discussed in #7315
2. Skip the permissions check for compiled/standalone binaries (i.e. [`is_standalone_binary()`](https://github.com/denoland/deno/blob/v1.35.2/cli/standalone/binary.rs#L211))
3. Only perform permissions check if the `to_file_path()` returns an OK result
    * i.e. `if let Ok(main_file_path = main_url.to_file_path()) { /* Perform permission check */ }`)

Option 1 (#7315) seems like the best option as this has been previously discussed and would remove the need to convert the URL to a platform-specific file path.

**Comment by jeff-hykin** [comment]
Took me a while to debug, but yep I'm affected by this as well. If `Deno.mainModule` threw an error instead of crashing that would be really helpful.

**Comment by jeff-hykin** [comment]
Running this before any imports is a successful workaround for me. However, to make this code run before any imports  it needs to a sync import above all the others.

```js
function fromFileUrl(url) {
    url = url instanceof URL ? url : new URL(url);
    if (Deno.build.os == "windows") {
        if (url.protocol != "file:") {
            throw new TypeError("Must be a file URL.");
        }
        let path7 = decodeURIComponent(
            url.pathname.replace(/\//g, "\\").replace(/%(?![0-9A-Fa-f]{2})/g, "%25")
        ).replace(/^\\*([A-Za-z]:)(\\|$)/, "$1\\");
        if (url.hostname != "") {
            path7 = `\\\\${url.hostname}${path7}`;
        }
        return path7;
    } else {
        if (url.protocol != "file:") {
            throw new TypeError("Must be a file URL.");
        }
        return decodeURIComponent(
            url.pathname.replace(/%(?![0-9A-Fa-f]{2})/g, "%25")
        );
    }
}
Object.defineProperty(Deno, "mainModule", {
    get() {
        const err = new Error()
        const filePaths = [...err.stack.matchAll(/^.+?(file:\/\/\/[\w\W]*?):/gm)].map(each=>each[1])
        const lastPath = filePaths.slice(-1)[0];
            if (lastPath) {
                try {
                    if (Deno.statSync(fromFileUrl(lastPath)).isFile) {
                        return lastPath;
                    }
                } catch (error) {
            }
        }
    }
})
```

This can be done inline with this import:

```js
import "data:text/javascript;base64,ICBmdW5jdGlvbiBmcm9tRmlsZVVybCh1cmwpIHsKICAgIHVybCA9IHVybCBpbnN0YW5jZW9mIFVSTCA/IHVybCA6IG5ldyBVUkwodXJsKTsKICAgIGlmIChEZW5vLmJ1aWxkLm9zID09ICJ3aW5kb3dzIikgewogICAgICBpZiAodXJsLnByb3RvY29sICE9ICJmaWxlOiIpIHsKICAgICAgICB0aHJvdyBuZXcgVHlwZUVycm9yKCJNdXN0IGJlIGEgZmlsZSBVUkwuIik7CiAgICAgIH0KICAgICAgbGV0IHBhdGg3ID0gZGVjb2RlVVJJQ29tcG9uZW50KHVybC5wYXRobmFtZS5yZXBsYWNlKC9cLy9nLCAiXFwiKS5yZXBsYWNlKC8lKD8hWzAtOUEtRmEtZl17Mn0pL2csICIlMjUiKSkucmVwbGFjZSgvXlxcKihbQS1aYS16XTopKFxcfCQpLywgIiQxXFwiKTsKICAgICAgaWYgKHVybC5ob3N0bmFtZSAhPSAiIikgewogICAgICAgIHBhdGg3ID0gYFxcXFwke3VybC5ob3N0bmFtZX0ke3BhdGg3fWA7CiAgICAgIH0KICAgICAgcmV0dXJuIHBhdGg3OwogICAgfSBlbHNlIHsKICAgICAgaWYgKHVybC5wcm90b2NvbCAhPSAiZmlsZToiKSB7CiAgICAgICAgdGhyb3cgbmV3IFR5cGVFcnJvcigiTXVzdCBiZSBhIGZpbGUgVVJMLiIpOwogICAgICB9CiAgICAgIHJldHVybiBkZWNvZGVVUklDb21wb25lbnQodXJsLnBhdGhuYW1lLnJlcGxhY2UoLyUoPyFbMC05QS1GYS1mXXsyfSkvZywgIiUyNSIpKTsKICAgIH0KICB9CiAgT2JqZWN0LmRlZmluZVByb3BlcnR5KERlbm8sICJtYWluTW9kdWxlIiwgewogICAgZ2V0ICgpIHsKICAgICAgY29uc3QgZXJyID0gbmV3IEVycm9yKCk7CiAgICAgIGNvbnN0IGZpbGVQYXRocyA9IFsKICAgICAgICAuLi5lcnIuc3RhY2subWF0Y2hBbGwoL14uKz8oZmlsZTpcL1wvXC9bXHdcV10qPyk6L2dtKQogICAgICBdLm1hcCgoZWFjaCk9PmVhY2hbMV0pOwogICAgICBjb25zdCBsYXN0UGF0aCA9IGZpbGVQYXRocy5zbGljZSgtMSlbMF07CiAgICAgIGlmIChsYXN0UGF0aCkgewogICAgICAgIHRyeSB7CiAgICAgICAgICBpZiAoRGVuby5zdGF0U3luYyhmcm9tRmlsZVVybChsYXN0UGF0aCkpLmlzRmlsZSkgewogICAgICAgICAgICByZXR1cm4gbGFzdFBhdGg7CiAgICAgICAgICB9CiAgICAgICAgfSBjYXRjaCAoZXJyb3IpIHt9CiAgICAgIH0KICAgIH0KICB9KTs="
```

And you can generate that import with the following deno code:
```js
function patch() {
    function fromFileUrl(url) {
        url = url instanceof URL ? url : new URL(url);
        if (Deno.build.os == "windows") {
            if (url.protocol != "file:") {
                throw new TypeError("Must be a file URL.");
            }
            let path7 = decodeURIComponent(
                url.pathname.replace(/\//g, "\\").replace(/%(?![0-9A-Fa-f]{2})/g, "%25")
            ).replace(/^\\*([A-Za-z]:)(\\|$)/, "$1\\");
            if (url.hostname != "") {
                path7 = `\\\\${url.hostname}${path7}`;
            }
            return path7;
        } else {
            if (url.protocol != "file:") {
                throw new TypeError("Must be a file URL.");
            }
            return decodeURIComponent(
                url.pathname.replace(/%(?![0-9A-Fa-f]{2})/g, "%25")
            );
        }
    }
    Object.defineProperty(Deno, "mainModule", {
        get() {
            const err = new Error()
            const filePaths = [...err.stack.matchAll(/^.+?(file:\/\/\/[\w\W]*?):/gm)].map(each=>each[1])
            const lastPath = filePaths.slice(-1)[0];
                if (lastPath) {
                    try {
                        if (Deno.statSync(fromFileUrl(lastPath)).isFile) {
                            return lastPath;
                        }
                    } catch (error) {
                }
            }
        }
    })
}
var codeContent = patch.toString().slice("function patch() {\n".length, -2)
console.log(`import "data:text/javascript;base64,${btoa(codeContent)}"`)
```

**Comment by arite** [comment]
Fixed by #25667.

I have re-tested this issue using Deno `v2.0.0` and the issue is no longer present/looks to have been fixed.
